### PR TITLE
[core] add operation tracing via Apache HTrace.

### PR DESCRIPTION
**Pull request by busbey**
add basic htrace instrumentation points and configs for turning htrace on/off. defaults to off.

when merged, we should provide some sample docs on using this to solve e.g. #415 .

### CODE REVIEW

File: core/src/main/java/com/yahoo/ycsb/TerminatorThread.java [hunk 1]
```diff
@@ -16,12 +16,11 @@
...
 package com.yahoo.ycsb;
 
-import java.util.List;
-import java.util.Vector;
+import java.util.Collection;
 
 /**
  * A thread that waits for the maximum specified time and then interrupts all the client
- * threads passed as the Vector at initialization of this thread.
+ * threads passed  at initialization of this thread.
```
Comment (cmccoy) [inline core/src/main/java/com/yahoo/ycsb/TerminatorThread.java]: Crazy nit: extra space here.

**Comment by cmccoy** [comment]
+1. Is there any non-negligible overhead with tracing disabled?

**Comment by busbey** [comment]
I haven't verified the overhead experimentally yet. Looking at the code for the HTrace library, when things are disabled I think it just constantly returns a singleton.

**Comment by cmccoy** [comment]
Sounds good to me.

**Comment by gkamat** [comment]
I think it would be interesting to check out the amount of overhead HTrace imposes.  I'll run some tests with and without the change and then check performance after enabling tracing.  Could we hold off merging the change until I get some numbers?  Thanks.

**Comment by busbey** [comment]
I'm not in a hurry, wfm.

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 2]
```diff
@@ -469,6 +475,10 @@ public int getOpsTodo()
    */
   public static final String MAX_EXECUTION_TIME = "maxexecutiontime";
 
+  /**
+   * All keys for configuring the tracing system start with this prefix.
+   */
+  private static final String HTRACE_KEY_PREFIX="htrace.";
```
Comment (stfeng) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: consider renaming this to "htrace.ycsb_core", I can see in the future some DB binding may be interested in using htrace too.

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 3]
```diff
@@ -801,67 +820,76 @@ public void run()
...
+    try (final TraceScope span = tracer.newScope("Client#workload_init")) {
+
+      try
+      {
+        workload.init(props);
+      }
+      catch (WorkloadException e)
+      {
+        e.printStackTrace();
+        e.printStackTrace(System.out);
```
Comment (stfeng) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: Is there a need to flush the "tracer" at here before System.exit(0)? The reason I ask is I'm not sure how the automatically scoped TraceScope works with abnormal exit like System.exit(), i.e., whether span.finalize() would be called in the event of System.ext()... (My understanding is the process is terminating, there is no need to do GC and reclaim out of scope variables anymore, so the flush won't happen...).

Or we can move System.exit(0) outside the scope of TraceScope.

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 4]
```diff
@@ -878,64 +906,79 @@ public void run()
...
+
+      try
+      {
+        workload.cleanup();
+      }
+      catch (WorkloadException e)
+      {
+        e.printStackTrace();
+        e.printStackTrace(System.out);
+        System.exit(0);
```
Comment (stfeng) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: same comment as above.

**Comment by stfeng** [comment]
One more (minor) comment: it'd be nice if you could break this commit into two, one for the formatting changes only, and the second one contains new logic. It makes change isolation easier...

But it's up to you, I don't have a strong opinion on this (?w=0 works well for me as a reviewer ;)

File: core/src/main/java/com/yahoo/ycsb/Client.java (hunk 3 above)
Comment (busbey) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: this is a very good point. I know {{finally}} blocks aren't used in the case of System.exit, but I don't know about try-with-resources.

In any case, I'll move the Sysetm.exit out or replace it.

**Comment by busbey** [comment]
AFAIK, the only formatting changes are from when I'm placing lines into a try block (and thus their indentation level should change).

File: core/src/main/java/com/yahoo/ycsb/Client.java (hunk 3 above)
Comment (saintstack) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: No harm closing the Tracer when done. This should flush any accumulated Traces as long as the process is up. There is no #flush on Tracer. Maybe there should be.

Comment (cmccabe) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: HTrace registers a shutdown hook with java.lang.Runtime#addShutdownHook.  This hook closes all SpanReceiver objects when the JVM exits.  SpanReceiver#close will attempt to persist the spans to htraced (or wherever that particular SpanReceiver sends spans).  In the case of HTracedSpanReceiver it will try to send them to htraced.

This shutdown hook will be run in the System.exit(0) case you guys are talking about.  It is also run in the SIGTERM (i.e. user sends control-C or normal kill) case.  It's not run in SIGKILL, but that is by necessity.  SIGKILL _always_ kills the process immediately without giving it the chance to do any cleanup.

Bottom line: this is not an issue.  You don't need a finally block here, or to worry about flushing anything on shutdown.

Comment (stfeng) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: Thanks for the clarification guys, nice discussions on HTrace. 

@busbey - Code LGTM as is then.

**Comment by busbey** [comment]
rebased to current master, made sure we didn't have a System.exit within any of our htrace blocks.

@gkamat how're things looking?

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 5]
```diff
@@ -778,6 +796,9 @@ public void run()
       }
     };
 
+    Workload workload=null;
```
Comment (clehene) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: style: spaces around `=`

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 6]
```diff
@@ -801,9 +820,14 @@ public void run()
       System.exit(0);
     }
 
+    final Tracer tracer = new Tracer.Builder("YCSB " + workload.getClass().getSimpleName()).conf(conf).build();
```
Comment (clehene) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: style: these builder chains are easier to read when operations are broken one per line
e.g.
```
Tracer.Builder("YCSB " + workload.getClass().getSimpleName())
    .conf(conf)
    .build();
```

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 7]
```diff
@@ -801,9 +820,14 @@ public void run()
...
+
     try
     {
-      workload.init(props);
+      try (final TraceScope span = tracer.newScope("Client#workload_init")) {
```
Comment (clehene) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: Extract constant to be consistent with the rest of the code

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 8]
```diff
@@ -812,56 +836,65 @@ public void run()
...
-    int opcount;
-    if (dotransactions)
-    {
-      opcount=Integer.parseInt(props.getProperty(OPERATION_COUNT_PROPERTY,"0"));
-    }
-    else
-    {
-      if (props.containsKey(INSERT_COUNT_PROPERTY))
+    boolean initFailed = false;
+    try (final TraceScope span = tracer.newScope("Client#init")) {
```
Comment (clehene) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: Extract constant

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 9]
```diff
@@ -812,56 +836,65 @@ public void run()
...
-      catch (UnknownDBException e)
+      for (int threadid=0; threadid<threadcount; threadid++)
       {
-        System.out.println("Unknown DB "+dbname);
-        System.exit(0);
-      }
+        DB db=null;
+        try
+        {
+          db=DBFactory.newDB(dbname, props, tracer);
```
Comment (clehene) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: Style: spaces around `=`

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 10]
```diff
@@ -812,56 +836,65 @@ public void run()
...
+        }
+        catch (UnknownDBException e)
+        {
+          System.out.println("Unknown DB "+dbname);
```
Comment (clehene) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: Style: spaces around `+`

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 11]
```diff
@@ -812,56 +836,65 @@ public void run()
...
+          initFailed = true;
+          break;
+        }
 
 
-      int threadopcount = opcount/threadcount;
+        int threadopcount = opcount/threadcount;
```
Comment (clehene) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: Style; spaces around `/`

File: core/src/main/java/com/yahoo/ycsb/Client.java [hunk 12]
```diff
@@ -878,54 +911,69 @@ public void run()
...
     Thread terminator = null;
+    long st;
+    long en;
+    int opsDone;
 
-    if (maxExecutionTime > 0) {
-      terminator = new TerminatorThread(maxExecutionTime, clients, workload);
-      terminator.start();
-    }
+    try (final TraceScope span = tracer.newScope("Client#workload")) {
```
Comment (clehene) [inline core/src/main/java/com/yahoo/ycsb/Client.java]: extract constant

**Comment by busbey** [comment]
Rebased to current master. incorporated feedback from @clehene.

Thanks for the review!

**Comment by saintstack** [comment]
I ran a basic compare of before and after applying the patch at 0f674fc7278c947acc25a596ef4a7d97ca8800c1

Client is doing ycsb workload c with 48 threads doing 100M ops.

My first test was running the client as part of a fleet of 8 nodes. In this case it the patched (but w/o htrace enabled) ran faster/did less work.

This is without the patch:
```

2016-04-28 13:26:28:617 679 sec: 10000000 operations; 889.26 current ops/sec; [CLEANUP: Count=36, Max=44191, Min=2, Avg=1236.33, 90=20, 99=44191, 99.9=44191, 99.99=44191] [READ: Count=8181, Max=634367, Min=334, Avg=3065, 90=4879, 99=16271, 99.9=209279, 99.99=592383]

Performance counter stats for '/home/stack/ycsb-0.9.0-SNAPSHOT/bin/ycsb run hbase10 -P /home/stack/ycsb-0.9.0-SNAPSHOT/workloads/workloadc -p table=ycsb -cp /home/stack/conf_hbase -p columnfamily=family -s -threads 48 -jvm-args=-Xmx8192m -Djava.security.egd=file:/dev/./urandom  -p recordcount=10000000 -p operati    oncount=10000000 -p exportfile=logs/ycsb-workloadc-measurements-ve0524-20160428T20:15:08Z.json -p exporter=com.yahoo.ycsb.measurements.exporter.JSONArrayMeasurementsExporter':
    1633454.192658 task-clock (msec)         #    2.401 CPUs utilized
        26,310,835 context-switches          #    0.016 M/sec
         4,688,282 cpu-migrations            #    0.003 M/sec
         1,265,819 page-faults               #    0.775 K/sec
 2,082,072,847,238 cycles                    #    1.275 GHz
   <not supported> stalled-cycles-frontend
   <not supported> stalled-cycles-backend
 1,071,649,307,268 instructions              #    0.51  insns per cycle
   204,398,225,103 branches                  #  125.133 M/sec
     8,028,409,082 branch-misses             #    3.93% of all branches

     680.266738512 seconds time elapsed
```

This is with the patch:

```
2016-04-28 13:40:47:499 531 sec: 10000000 operations; 1097.73 current ops/sec; [CLEANUP: Count=10, Max=37631, Min=2, Avg=3767.6, 90=14, 99=37631, 99.9=37631, 99.99=37631] [READ: Count=1439, Max=91007, Min=307, Avg=1281.14, 90=2187, 99=10471, 99.9=46239, 99.99=91007]

 Performance counter stats for '/home/stack/ycsb-0.9.0-SNAPSHOT/bin/ycsb run hbase10 -P /home/stack/ycsb-0.9.0-SNAPSHOT/workloads/workloadc -p table=ycsb -cp /home/stack/conf_hbase -p columnfamily=family -s -threads 48 -jvm-args=-Xmx8192m -Djava.security.egd=file:/dev/./urandom  -p recordcount=10000000 -p operationcount=10000000 -p exportfile=logs/ycsb-workloadc-measurements-ve0524-20160428T20:31:55Z.json -p exporter=com.yahoo.ycsb.measurements.exporter.JSONArrayMeasurementsExporter':

    1603640.139671 task-clock (msec)         #    3.012 CPUs utilized
        26,357,262 context-switches          #    0.016 M/sec
         4,693,836 cpu-migrations            #    0.003 M/sec
         1,432,836 page-faults               #    0.893 K/sec
 2,045,935,113,904 cycles                    #    1.276 GHz
   <not supported> stalled-cycles-frontend
   <not supported> stalled-cycles-backend
 1,056,620,914,643 instructions              #    0.52  insns per cycle
   199,977,636,377 branches                  #  124.702 M/sec
     7,880,502,259 branch-misses             #    3.94% of all branches

     532.391661177 seconds time elapsed
```

Let me try the patched client on its ownsome and run it a few times....

**Comment by saintstack** [comment]
Ok. I ran with long client running 100M ops with 48 threads before and after applying the patch. Did it three times. With the patch, we might be doing a few more instructions but about same instructions per cycle and same overall throughput roughly

Without patch
```

 2016-04-28 14:04:57,354 INFO  [Thread-2-EventThread] zookeeper.ClientCnxn: EventThread shut down
 2016-04-28 14:04:57:456 127 sec: 10000000 operations; 75348.03 current ops/sec; [READ: Count=582518, Max=35007, Min=234, Avg=596.31, 90=707, 99=7623, 99.9=14375, 99.99=20367] [CLEANUP: Count=96, Max=124671, Min=2, Avg=1317.21, 90=22, 99=785, 99.9=124671, 99.99=124671]

  Performance counter stats for '/home/stack/ycsb-0.9.0-SNAPSHOT/bin/ycsb run hbase10 -P /home/stack/ycsb-0.9.0-SNAPSHOT/workloads/workloadc -p table=ycsb -cp /home/stack/conf_hbase -p columnfamily=family -s -threads 48 -jvm-args=-Xmx8192m -Djava.security.egd=file:/dev/./urandom  -p
 recordcount=10000000 -p operati    oncount=10000000 -p exportfile=logs/ycsb-workloadc-measurements-ve0524-20160428T21:02:48Z.json -p exporter=com.yahoo.ycsb.measurements.exporter.JSONArrayMeasurementsExporter':

     1448443.518777 task-clock (msec)         #   11.264 CPUs utilized
         28,226,146 context-switches          #    0.019 M/sec
         10,228,545 cpu-migrations            #    0.007 M/sec
            729,752 page-faults               #    0.504 K/sec
  2,016,484,967,472 cycles                    #    1.392 GHz
    <not supported> stalled-cycles-frontend
    <not supported> stalled-cycles-backend
  1,003,193,720,859 instructions              #    0.50  insns per cycle
    178,052,034,558 branches                  #  122.926 M/sec
      2,544,438,220 branch-misses             #    1.43% of all branches

      128.591543660 seconds time elapsed

...
 2016-04-28 14:27:38:492 127 sec: 10000000 operations; 75753.02 current ops/sec; [READ: Count=589813, Max=36639, Min=232, Avg=601.35, 90=725, 99=8091, 99.9=12559, 99.99=18671] [CLEANUP: Count=96, Max=122879, Min=2, Avg=1296.81, 90=18, 99=770, 99.9=122879, 99.99=122879]

  Performance counter stats for '/home/stack/ycsb-0.9.0-SNAPSHOT/bin/ycsb run hbase10 -P /home/stack/ycsb-0.9.0-SNAPSHOT/workloads/workloadc -p table=ycsb -cp /home/stack/conf_hbase -p columnfamily=family -s -threads 48 -jvm-args=-Xmx8192m -Djava.security.egd=file:/dev/./urandom  -p
 recordcount=10000000 -p operati    oncount=10000000 -p exportfile=logs/ycsb-workloadc-measurements-ve0524-20160428T21:25:30Z.json -p exporter=com.yahoo.ycsb.measurements.exporter.JSONArrayMeasurementsExporter':

     1429560.096961 task-clock (msec)         #   11.119 CPUs utilized
         28,209,708 context-switches          #    0.020 M/sec
         10,189,884 cpu-migrations            #    0.007 M/sec
            570,630 page-faults               #    0.399 K/sec
  1,976,595,469,731 cycles                    #    1.383 GHz
    <not supported> stalled-cycles-frontend
    <not supported> stalled-cycles-backend
  1,000,725,835,913 instructions              #    0.51  insns per cycle
    176,393,141,848 branches                  #  123.390 M/sec
      2,517,562,830 branch-misses             #    1.43% of all branches

      128.573300989 seconds time elapsed

...
 2016-04-28 14:48:17,500 DEBUG [Thread-28] ipc.AbstractRpcClient: Stopping rpc client
 2016-04-28 14:48:17:602 130 sec: 10000000 operations; 17352.94 current ops/sec; [READ: Count=8769, Max=12607, Min=253, Avg=466.81, 90=530, 99=1250, 99.9=10519, 99.99=11831] [CLEANUP: Count=66, Max=121407, Min=2, Avg=1846.58, 90=13, 99=34, 99.9=121407, 99.99=121407]

  Performance counter stats for '/home/stack/ycsb-0.9.0-SNAPSHOT/bin/ycsb run hbase10 -P /home/stack/ycsb-0.9.0-SNAPSHOT/workloads/workloadc -p table=ycsb -cp /home/stack/conf_hbase -p columnfamily=family -s -threads 48 -jvm-args=-Xmx8192m -Djava.security.egd=file:/dev/./urandom  -p
 recordcount=10000000 -p operati    oncount=10000000 -p exportfile=logs/ycsb-workloadc-measurements-ve0524-20160428T21:46:06Z.json -p exporter=com.yahoo.ycsb.measurements.exporter.JSONArrayMeasurementsExporter':

     1462104.055059 task-clock (msec)         #   11.132 CPUs utilized
         28,187,645 context-switches          #    0.019 M/sec
         10,214,228 cpu-migrations            #    0.007 M/sec
            656,063 page-faults               #    0.449 K/sec
  2,032,209,749,245 cycles                    #    1.390 GHz
    <not supported> stalled-cycles-frontend
    <not supported> stalled-cycles-backend
  1,002,089,314,886 instructions              #    0.49  insns per cycle
    176,808,449,062 branches                  #  120.927 M/sec
      2,491,319,581 branch-misses             #    1.41% of all branches

      131.339745942 seconds time elapsed   
```

With patch
[
...
2016-04-28 13:50:37,676 INFO  [Thread-11-EventThread] zookeeper.ClientCnxn: EventThread shut down
2016-04-28 13:50:37:704 127 sec: 10000000 operations; 77356.75 current ops/sec; [READ: Count=563702, Max=24703, Min=231, Avg=585.2, 90=706, 99=7447, 99.9=12087, 99.99=19823] [CLEANUP: Count=96, Max=53023, Min=2, Avg=781.72, 90=22, 99=20527, 99.9=53023, 99.99=53023]

 Performance counter stats for '/home/stack/ycsb-0.9.0-SNAPSHOT/bin/ycsb run hbase10 -P /home/stack/ycsb-0.9.0-SNAPSHOT/workloads/workloadc -p table=ycsb -cp /home/stack/conf_hbase -p columnfamily=family -s -threads 48 -jvm-args=-Xmx8192m -Djava.security.egd=file:/dev/./urandom  -p recordcount=10000000 -p operationco
unt=10000000 -p exportfile=logs/ycsb-workloadc-measurements-ve0524-20160428T20:48:29Z.json -p exporter=com.yahoo.ycsb.measurements.exporter.JSONArrayMeasurementsExporter':

    1498744.776813 task-clock (msec)         #   11.699 CPUs utilized
        28,287,296 context-switches          #    0.019 M/sec
        10,320,958 cpu-migrations            #    0.007 M/sec
           732,257 page-faults               #    0.489 K/sec
 2,106,401,174,465 cycles                    #    1.405 GHz
   <not supported> stalled-cycles-frontend
   <not supported> stalled-cycles-backend
 1,014,689,743,992 instructions              #    0.48  insns per cycle
   179,404,941,364 branches                  #  119.703 M/sec
     2,607,635,152 branch-misses             #    1.45% of all branches

     128.111865634 seconds time elapsed


...
2016-04-28 13:55:31,779 DEBUG [Thread-49] ipc.AbstractRpcClient: Stopping rpc client
2016-04-28 13:55:31:803 125 sec: 10000000 operations; 74660 current ops/sec; [READ: Count=425694, Max=49183, Min=242, Avg=597.46, 90=718, 99=7127, 99.9=14751, 99.99=47775] [CLEANUP: Count=96, Max=45151, Min=1, Avg=487.41, 90=19, 99=864, 99.9=45151, 99.99=45151]

 Performance counter stats for '/home/stack/ycsb-0.9.0-SNAPSHOT/bin/ycsb run hbase10 -P /home/stack/ycsb-0.9.0-SNAPSHOT/workloads/workloadc -p table=ycsb -cp /home/stack/conf_hbase -p columnfamily=family -s -threads 48 -jvm-args=-Xmx8192m -Djava.security.egd=file:/dev/./urandom  -p recordcount=10000000 -p operationco
unt=10000000 -p exportfile=logs/ycsb-workloadc-measurements-ve0524-20160428T20:53:25Z.json -p exporter=com.yahoo.ycsb.measurements.exporter.JSONArrayMeasurementsExporter':

    1428719.305351 task-clock (msec)         #   11.294 CPUs utilized
        28,211,365 context-switches          #    0.020 M/sec
        10,209,394 cpu-migrations            #    0.007 M/sec
           680,489 page-faults               #    0.476 K/sec
 1,988,149,877,105 cycles                    #    1.392 GHzG
   <not supported> stalled-cycles-frontend
   <not supported> stalled-cycles-backend
 1,003,002,120,666 instructions              #    0.50  insns per cycle
   177,087,226,485 branches                  #  123.948 M/sec
     2,510,017,572 branch-misses             #    1.42% of all branches

     126.506714064 seconds time elapsed

...
2016-04-28 13:59:03,343 INFO  [Thread-11-EventThread] zookeeper.ClientCnxn: EventThread shut down
2016-04-28 13:59:03,343 DEBUG [Thread-16] ipc.AbstractRpcClient: Stopping rpc client
2016-04-28 13:59:03:377 126 sec: 10000000 operations; 76331.57 current ops/sec; [READ: Count=476309, Max=25743, Min=234, Avg=589.26, 90=700, 99=4459, 99.9=14887, 99.99=22527] [CLEANUP: Count=96, Max=58303, Min=2, Avg=624.32, 90=22, 99=783, 99.9=58303, 99.99=58303]

 Performance counter stats for '/home/stack/ycsb-0.9.0-SNAPSHOT/bin/ycsb run hbase10 -P /home/stack/ycsb-0.9.0-SNAPSHOT/workloads/workloadc -p table=ycsb -cp /home/stack/conf_hbase -p columnfamily=family -s -threads 48 -jvm-args=-Xmx8192m -Djava.security.egd=file:/dev/./urandom  -p recordcount=10000000 -p operationco
unt=10000000 -p exportfile=logs/ycsb-workloadc-measurements-ve0524-20160428T20:56:56Z.json -p exporter=com.yahoo.ycsb.measurements.exporter.JSONArrayMeasurementsExporter':

    1486900.980098 task-clock (msec)         #   11.698 CPUs utilized
        28,250,518 context-switches          #    0.019 M/sec
        10,297,318 cpu-migrations            #    0.007 M/sec
           717,429 page-faults               #    0.482 K/sec
 2,086,691,492,394 cycles                    #    1.403 GHz
   <not supported> stalled-cycles-frontend
   <not supported> stalled-cycles-backend
 1,012,490,777,213 instructions              #    0.49  insns per cycle
   178,904,362,508 branches                  #  120.320 M/sec
     2,517,984,046 branch-misses             #    1.41% of all branches

     127.102766741 seconds time elapsed](url)

**Comment by saintstack** [comment]
Formatting is a bit messed up....

I'd say its fine to commit.

**Comment by busbey** [comment]
sounds good to me. I'll get a rebase on current master this weekend.

if anyone has other concerns they'd like handled please speak up quickly. Otherwise I'll move forward based on @cmccoy and @saintstack giving me +1s.

**Comment by busbey** [comment]
thanks for all the feedback / reviews!
